multidialer.go

```diff
@@ -24,87 +25,65 @@ func randAddr(addrs []*net.TCPAddr) *net.TCPAddr {
 	return nil
 }
 
+// DialContext dials a target addr.
+// Dialing preference is
+// * If there is a listener on the local interface the OS expects to use to route towards addr, use that.
+// * If there is a listener on a loopback address, addr is loopback, use that.
+// * If there is a listener on an undefined address (0.0.0.0 or ::), use that.
+// * Use the fallback IP specified during construction, with a port that's already being listened on, if one exists.
 func (d *multiDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
 	tcpAddr, err := net.ResolveTCPAddr(network, addr)
 	if err != nil {
 		return nil, err
 	}
+	ip := tcpAddr.IP
+	if !ip.IsLoopback() && !ip.IsGlobalUnicast() {
+		return nil, fmt.Errorf("undialable IP: %s", ip)
+	}
 
-	// We pick the source *port* based on the following algorithm.
-	//
-	// 1. If we're dialing loopback, choose a source-port in order of
-	//    preference:
-	//   1. A port in-use by an explicit loopback listener.
-	//   2. A port in-use by a listener on an unspecified address (must
-	//      also be listening on localhost).
-	//   3. A port in-use by a listener on a global address. We don't have
-	//      any other better options (other than picking a random port).
-	// 2. If we're dialing a global address, choose a source-port in order
-	//    of preference:
-	//   1. A port in-use by a listener on an unspecified address (the most
-	//      general case).
-	//   2. A port in-use by a listener on the global address.
-	// 3. Fail on link-local dials (go-ipfs currently forbids this and I
-	//    figured we could try lifting this restriction later).
-	//
-	//
-	// Note: We *always* dial from the unspecified address (regardless of
-	// the port we pick). In the future, we could use netlink (on Linux) to
-	// figure out the right source address but we're going to punt on that.
+	router, err := netroute.New()
+	if err != nil {
+		return nil, err
```

Comment:
I think it actually may make more sense to not fail the full dial if there are problems with netroute. There are other listeners / hints we can still use, so this shouldn't be a fatal issue.